File: Src/CPP/version_manager.cpp
#include "version_manager.hpp"
#include "http_client.hpp"

namespace VersionManager
{
    int compare(const std::string &v1, const std::string &v2)
    {
        std::vector<int> ver1, ver2;
        std::stringstream ss1(v1), ss2(v2);
        std::string segment;

        while (std::getline(ss1, segment, '.'))
        {
            ver1.push_back(std::stoi(segment));
        }
        while (std::getline(ss2, segment, '.'))
        {
            ver2.push_back(std::stoi(segment));
        }

        for (size_t i = 0; i < std::min(ver1.size(), ver2.size()); i++)
        {
            if (ver1[i] < ver2[i])
                return -1;
            if (ver1[i] > ver2[i])
                return 1;
        }

        return 0;
    }

    bool checkAndInstallPythonLibraries()
    {
        // 检查python是否可用
        int pythonCheck = system("python --version > /dev/null 2>&1");
        if (pythonCheck != 0)
        {
            std::println("\033[31mPython未安装或不可用\033[0m");
            return false;
        }

        // 检查google_play_scraper库是否已安装
        std::string checkScript = R"(import importlib.util
import sys
print("installed" if importlib.util.find_spec("google_play_scraper") else "not_installed")
)";

        // 检查脚本文件是否存在，不存在则创建
        if (!fs::exists("check_library.py"))
        {
            std::ofstream scriptFile("check_library.py");
            scriptFile << checkScript;
            scriptFile.close();
        }

        FILE *pipe = popen("python check_library.py", "r");
        if (!pipe)
        {
            return false;
        }

        char buffer[128];
        std::string result = "";
        while (!feof(pipe))
        {
            if (fgets(buffer, 128, pipe) != nullptr)
                result += buffer;
        }
        pclose(pipe);

        // 去除结果中的空白字符
        result.erase(std::remove_if(result.begin(), result.end(), ::isspace), result.end());

        if (result == "not_installed")
        {
            std::println("检测到缺失依赖，正在安装...");
            int installResult = system("pip install google_play_scraper --user > /dev/null 2>&1");
            if (installResult != 0)
            {
                std::println("\033[31m安装缺失依赖失败\033[0m");
                return false;
            }
            std::println("\033[32m成功安装缺失依赖\033[0m");
        }

        return true;
    }

    std::string getVersionFromGooglePlay()
    {
        std::string pythonScript = R"(import json
from google_play_scraper import app as playstore_app

try:
    result = playstore_app(app_id="com.kakaogames.eversoul", lang="en", country="kr")
    print(json.dumps({"success": True, "version": result["version"]}))
except Exception as e:
    print(json.dumps({"success": False, "error": str(e)}))
)";

        // 不存在则创建
        if (!fs::exists("version_scraper.py"))
        {
            std::ofstream scriptFile("version_scraper.py");
            scriptFile << pythonScript;
            scriptFile.close();
        }

        std::string command = "python version_scraper.py 2>/dev/null";
        FILE *pipe = popen(command.c_str(), "r");
        if (!pipe)
        {
            std::println("\033[31mPython脚本执行失败\033[0m");
            return "";
        }

        char buffer[128];
        std::string result = "";
        while (!feof(pipe))
        {
            if (fgets(buffer, 128, pipe) != nullptr)
                result += buffer;
        }
        pclose(pipe);

        try
        {
            json response = json::parse(result);
            if (response["success"])
            {
                return response["version"];
            }
            else
            {
                std::println("\033[33mPython获取版本号失败: {}\033[0m", response["error"].get<std::string>());
            }
        }
        catch (const std::exception &e)
        {
            std::println("\033[31mPython脚本返回结果解析失败: {}\033[0m", e.what());
        }

        return "";
    }

    std::vector<std::string> generatePossibleVersions(const std::string &baseVersion)
    {
        std::vector<std::string> versions;

        // 解析基础版本号
        std::regex version_regex(R"((\d+)\.(\d+)\.(\d+))");
        std::smatch matches;
        if (!std::regex_match(baseVersion, matches, version_regex))
        {
            return versions;
        }

        int major = std::stoi(matches[1]);
        int minor = std::stoi(matches[2]);
        int patch = std::stoi(matches[3]);

        // 当前minor版本，从当前patch开始，向上检查到200
        for (int p = patch; p <= 200; p++)
        {
            versions.push_back(std::format("{}.{}.{}", major, minor, p));
        }

        // 下一个minor版本，从0开始，向上检查到200
        for (int p = 0; p <= 200; p++)
        {
            versions.push_back(std::format("{}.{}.{}", major, minor + 1, p));
        }

        // major递增，minor从0开始，patch从0开始，向上检查到200
        for (int p = 0; p <= 200; p++)
        {
            versions.push_back(std::format("{}.{}.{}", major + 1, 0, p));
        }

        return versions;
    }

    CNServerConfig getCNServerConfig()
    {
        CNServerConfig config;
        config.isValid = false;
        config.reviewIsValid = false;

        try
        {
            std::string url = "http://yhlh-client.zlongame.com/YHLH/cbt/android/apps.json";
            std::string response = HttpClient::get(url);

            if (response.empty())
            {
                return config;
            }

            json data = json::parse(response);

            // === 获取 formal 正式服配置 ===
            // 从 formal.custom.patchversion 获取版本号
            if (data.contains("formal") &&
                data["formal"].contains("custom") &&
                data["formal"]["custom"].contains("patchversion"))
            {
                config.version = data["formal"]["custom"]["patchversion"].get<std::string>();
            }
            else
            {
                std::println("\033[31m国服formal配置中未找到版本号\033[0m");
            }

            // 从 formal.patch_url.bundle_down_url 获取下载URL列表
            if (data.contains("formal") &&
                data["formal"].contains("patch_url") &&
                data["formal"]["patch_url"].contains("bundle_down_url") &&
                data["formal"]["patch_url"]["bundle_down_url"].is_array())
            {
                for (const auto &url : data["formal"]["patch_url"]["bundle_down_url"])
                {
                    config.downloadUrls.push_back(url.get<std::string>());
                }
            }
            else
            {
                std::println("\033[31m国服formal配置中未找到下载URL\033[0m");
            }

            if (!config.downloadUrls.empty() && !config.version.empty())
            {
                config.isValid = true;
            }

            // === 获取 review 审核服配置 ===
            // 从 review.custom.patchversion 获取版本号
            if (data.contains("review") &&
                data["review"].contains("custom") &&
                data["review"]["custom"].contains("patchversion"))
            {
                config.reviewVersion = data["review"]["custom"]["patchversion"].get<std::string>();
                // std::println("获取到国服review版本号: \033[36m{}\033[0m", config.reviewVersion);
            }

            // 从 review.patch_url.bundle_down_url 获取下载URL列表
            if (data.contains("review") &&
                data["review"].contains("patch_url") &&
                data["review"]["patch_url"].contains("bundle_down_url") &&
                data["review"]["patch_url"]["bundle_down_url"].is_array())
            {
                for (const auto &url : data["review"]["patch_url"]["bundle_down_url"])
                {
                    config.reviewDownloadUrls.push_back(url.get<std::string>());
                }
            }

            if (!config.reviewDownloadUrls.empty() && !config.reviewVersion.empty())
            {
                config.reviewIsValid = true;
            }

            return config;
        }
        catch (const json::exception &e)
        {
            std::println("\033[31m解析国服配置失败: {}\033[0m", e.what());
            return config;
        }
        catch (const std::exception &e)
        {
            std::println("\033[31m获取国服配置失败: {}\033[0m", e.what());
            return config;
        }
    }
}

File: Src/Python/version_manager.py
"""
版本管理模块
"""

import importlib.util
import subprocess
import json
import re
from typing import Optional
from dataclasses import dataclass

from rich.console import Console
from http_client import HttpClient

console = Console()


@dataclass
class CNServerConfig:
    """国服配置信息结构"""

    version: str = ""
    download_urls: list[str] = None
    is_valid: bool = False
    
    # Review环境配置
    review_version: str = ""
    review_download_urls: list[str] = None
    review_is_valid: bool = False

    def __post_init__(self):
        if self.download_urls is None:
            self.download_urls = []
        if self.review_download_urls is None:
            self.review_download_urls = []


class VersionManager:
    """版本管理类"""

    @staticmethod
    def compare(v1: str, v2: str) -> int:
        """
        比较两个版本号

        Args:
            v1: 第一个版本号
            v2: 第二个版本号

        Returns:
            如果v1 < v2返回-1，如果v1 > v2返回1，如果v1 == v2返回0
        """
        ver1 = [int(x) for x in v1.split(".")]
        ver2 = [int(x) for x in v2.split(".")]

        for i in range(min(len(ver1), len(ver2))):
            if ver1[i] < ver2[i]:
                return -1
            if ver1[i] > ver2[i]:
                return 1

        return 0

    @staticmethod
    def check_and_install_python_libraries() -> bool:
        """
        检查并安装必要的Python库

        Returns:
            如果Python环境正确且所需库已安装或成功安装则返回True，否则返回False
        """
        # 检查google_play_scraper库是否已安装
        spec = importlib.util.find_spec("google_play_scraper")

        if spec is None:
            console.print("检测到缺失依赖，正在安装...")
            try:
                result = subprocess.run(
                    ["pip", "install", "google_play_scraper", "--user"],
                    capture_output=True,
                    text=True,
                    timeout=60,
                )
                if result.returncode != 0:
                    console.print("[bold red]安装缺失依赖失败[/bold red]")
                    return False
                console.print("[bold green]成功安装缺失依赖[/bold green]")
            except Exception as e:
                console.print(f"[bold red]安装依赖时出错: {e}[/bold red]")
                return False

        return True

    @staticmethod
    def get_version_from_google_play() -> str:
        """
        使用Python脚本从Google Play获取最新的应用版本号

        Returns:
            成功时返回版本号字符串，失败时返回空字符串
        """
        try:
            from google_play_scraper import app as playstore_app

            result = playstore_app(
                app_id="com.kakaogames.eversoul", lang="en", country="kr"
            )
            version = result.get("version", "")
            if version:
                console.print(f"获取到版本号: [bold cyan]{version}[/bold cyan]")
            return version
        except Exception as e:
            console.print(f"[bold yellow]Python获取版本号失败: {e}[/bold yellow]")
            return ""

    @staticmethod
    def generate_possible_versions(base_version: str) -> list[str]:
        """
        生成一系列可能的版本号（用于检查Review服务器）

        Args:
            base_version: 基础版本号

        Returns:
            包含所有可能版本号的列表
        """
        versions = []

        # 解析基础版本号
        match = re.match(r"(\d+)\.(\d+)\.(\d+)", base_version)
        if not match:
            return versions

        major = int(match.group(1))
        minor = int(match.group(2))
        patch = int(match.group(3))

        # 当前minor版本，从当前patch开始，向上检查到200
        for p in range(patch, 201):
            versions.append(f"{major}.{minor}.{p}")

        # 下一个minor版本，从0开始，向上检查到200
        for p in range(0, 201):
            versions.append(f"{major}.{minor + 1}.{p}")

        # major递增，minor从0开始，patch从0开始，向上检查到200
        for p in range(0, 201):
            versions.append(f"{major + 1}.0.{p}")

        return versions

    @staticmethod
    def get_cn_server_config() -> CNServerConfig:
        """
        从国服apps.json获取配置信息（包括formal和review环境）

        Returns:
            包含版本号和下载URL的配置结构
        """
        config = CNServerConfig()

        try:
            url = "http://yhlh-client.zlongame.com/YHLH/cbt/android/apps.json"
            response = HttpClient.get(url)

            if not response:
                return config

            data = json.loads(response)

            # === 获取 formal 正式服配置 ===
            # 从 formal.custom.patchversion 获取版本号
            if (
                "formal" in data
                and "custom" in data["formal"]
                and "patchversion" in data["formal"]["custom"]
            ):
                config.version = data["formal"]["custom"]["patchversion"]
            else:
                console.print("[bold red]国服formal配置中未找到版本号[/bold red]")

            # 从 formal.patch_url.bundle_down_url 获取下载URL列表
            if (
                "formal" in data
                and "patch_url" in data["formal"]
                and "bundle_down_url" in data["formal"]["patch_url"]
                and isinstance(data["formal"]["patch_url"]["bundle_down_url"], list)
            ):
                config.download_urls = data["formal"]["patch_url"]["bundle_down_url"]
            else:
                console.print("[bold red]国服formal配置中未找到下载URL[/bold red]")

            if config.download_urls and config.version:
                config.is_valid = True

            # === 获取 review 审核服配置 ===
            # 从 review.custom.patchversion 获取版本号
            if (
                "review" in data
                and "custom" in data["review"]
                and "patchversion" in data["review"]["custom"]
            ):
                config.review_version = data["review"]["custom"]["patchversion"]
                # console.print(f"获取到国服review版本号: [bold cyan]{config.review_version}[/bold cyan]")

            # 从 review.patch_url.bundle_down_url 获取下载URL列表
            if (
                "review" in data
                and "patch_url" in data["review"]
                and "bundle_down_url" in data["review"]["patch_url"]
                and isinstance(data["review"]["patch_url"]["bundle_down_url"], list)
            ):
                config.review_download_urls = data["review"]["patch_url"]["bundle_down_url"]

            if config.review_download_urls and config.review_version:
                config.review_is_valid = True

            return config
        except json.JSONDecodeError as e:
            console.print(f"[bold red]解析国服配置失败: {e}[/bold red]")
            return config
        except Exception as e:
            console.print(f"[bold red]获取国服配置失败: {e}[/bold red]")
            return config
